File: frontend/src/lib/browserCall.ts
// The dispatch conversation held here in the browser, instead of over the phone.
//
// The visitor plays the authority: Raqeeb speaks the same Arabic briefing it reads down a phone
// line, hears their reply through the microphone, and records their decision with the same tool.
// Audio goes straight from this page to OpenAI over WebRTC; Raqeeb's server only mints the
// short-lived client secret and takes back the result.
//
// The API key never reaches this file. The secret it uses is minted per incident, carries that
// incident's instructions, expires by itself, and is refused a second time by the backend.

const CALLS_URL = "https://api.openai.com/v1/realtime/calls"
const TOOL_NAME = "record_dispatch_confirmation"

// Returned to the model when it records a decision before anyone on the line has said a word,
// which it does: seen on a live call, it recorded a confirmation while the other side was still
// silent. Mirrors _NO_DECISION_HEARD in agent/voice/sip_authority_call.py, which holds the phone
// path to the same rule.
const NO_DECISION_HEARD =
  "لم يُسجَّل شيء: لم يتكلّم أحد من الجهة بعد، فلا يوجد قرار. " +
  "تأكّد أن من على الخط شخص من الجهة، وأكمل البلاغ، ولا تستدعِ الأداة إلا بعد أن تسمع قرارهم."

export type CallLine = { id: string; role: "assistant" | "authority"; text: string; final: boolean }

export type CallDecision = { confirmed: boolean; statement: string }

export type BrowserCallHandlers = {
  /** a line as it is being spoken; the same id arrives again as it grows */
  onLine: (line: CallLine) => void
  /** the authority's decision, the moment the model records it */
  onDecision: (decision: CallDecision) => void
  /** the conversation is over, for any reason */
  onEnd: (reason: "ended" | "timeout" | "error") => void
}

export type BrowserCall = {
  /** hang up; safe to call twice */
  stop: () => void
}

export class MicrophoneDenied extends Error {}

type Line = { role: "assistant" | "authority"; text: string; final: boolean }

/** Opens the microphone and connects it to the incident's conversation. */
export async function startBrowserCall(
  clientSecret: string,
  maxSeconds: number,
  handlers: BrowserCallHandlers,
): Promise<BrowserCall> {
  let microphone: MediaStream
  try {
    microphone = await navigator.mediaDevices.getUserMedia({ audio: true })
  } catch (error) {
    throw new MicrophoneDenied(String(error))
  }

  const connection = new RTCPeerConnection()
  const speaker = document.createElement("audio")
  speaker.autoplay = true
  const lines = new Map<string, Line>()
  let stopped = false
  let timer = 0

  const publish = (id: string, role: "assistant" | "authority", update: (line: Line) => void) => {
    const line = lines.get(id) ?? { role, text: "", final: false }
    update(line)
    lines.set(id, line)
    handlers.onLine({ id, ...line })
  }

  const stop = (reason: "ended" | "timeout" | "error" = "ended") => {
    if (stopped) return
    stopped = true
    window.clearTimeout(timer)
    for (const line of lines.values()) line.final = true
    microphone.getTracks().forEach((track) => track.stop())
    try {
      connection.close()
    } catch {
      // already closed
    }
    speaker.srcObject = null
    handlers.onEnd(reason)
  }

  connection.ontrack = (event) => {
    speaker.srcObject = event.streams[0]
  }
  connection.addTrack(microphone.getAudioTracks()[0], microphone)
  // the other party hanging up, or the network dropping, ends the conversation here too
  connection.onconnectionstatechange = () => {
    if (["failed", "closed", "disconnected"].includes(connection.connectionState)) stop("ended")
  }

  const channel = connection.createDataChannel("oai-events")
  const answerTool = (callId: string, output: Record<string, unknown>) => {
    channel.send(
      JSON.stringify({
        type: "conversation.item.create",
        item: { type: "function_call_output", call_id: callId, output: JSON.stringify(output) },
      }),
    )
    channel.send(JSON.stringify({ type: "response.create" }))
  }
  channel.addEventListener("message", (event) => {
    let message: Record<string, any>
    try {
      message = JSON.parse(event.data)
    } catch {
      return
    }
    if (import.meta.env.DEV) console.debug("realtime", message.type)
    switch (message.type) {
      // the authority's turn, claimed when it is committed so it keeps its place in the
      // conversation even though its transcription arrives later
      case "input_audio_buffer.committed":
        publish(message.item_id, "authority", () => {})
        break
      case "conversation.item.input_audio_transcription.delta":
        publish(message.item_id, "authority", (line) => (line.text += message.delta ?? ""))
        break
      case "conversation.item.input_audio_transcription.completed":
        publish(message.item_id, "authority", (line) => {
          if (message.transcript) line.text = message.transcript
          line.final = true
        })
        break
      case "response.output_audio_transcript.delta":
        publish(message.item_id, "assistant", (line) => (line.text += message.delta ?? ""))
        break
      case "response.output_audio_transcript.done":
        publish(message.item_id, "assistant", (line) => {
          if (message.transcript) line.text = message.transcript
          line.final = true
        })
        break
      case "response.function_call_arguments.done": {
        if (message.name !== TOOL_NAME) break
        if (![...lines.values()].some((line) => line.role === "authority")) {
          // a decision nobody has spoken is not one: refuse it and let the call carry on
          answerTool(message.call_id, { recorded: false, reason: NO_DECISION_HEARD })
          break
        }
        let args: { confirmed?: boolean; statement?: string } = {}
        try {
          args = JSON.parse(message.arguments || "{}")
        } catch {
          // a decision we cannot read is no decision
        }
        handlers.onDecision({ confirmed: Boolean(args.confirmed), statement: String(args.statement ?? "") })
        // let the call close politely, exactly as the phone path does, rather than cutting the
        // line the instant a decision lands
        answerTool(message.call_id, { acknowledged: true })
        break
      }
      case "error":
        console.error("realtime error", message.error)
        break
    }
  })
  // an outbound call: Raqeeb speaks first rather than waiting for a greeting that never comes
  channel.addEventListener("open", () => channel.send(JSON.stringify({ type: "response.create" })))

  try {
    const offer = await connection.createOffer()
    await connection.setLocalDescription(offer)
    const answer = await fetch(CALLS_URL, {
      method: "POST",
      body: offer.sdp,
      headers: { Authorization: `Bearer ${clientSecret}`, "Content-Type": "application/sdp" },
    })
    if (!answer.ok) throw new Error(`realtime call refused: ${answer.status}`)
    await connection.setRemoteDescription({ type: "answer", sdp: await answer.text() })
  } catch (error) {
    stop("error")
    throw error
  }

  // a ceiling on one conversation, so an abandoned tab cannot hold a session open
  timer = window.setTimeout(() => stop("timeout"), maxSeconds * 1000)
  return { stop: () => stop("ended") }
}

File: agent/routes/browser_call.py
"""The dispatch conversation held in the visitor's own browser, instead of over the phone.

The public dashboard has no telephony behind it: when an incident reaches the dispatch step, the
visitor's browser opens a WebRTC session straight to OpenAI's Realtime API and plays the part the
authority plays on the phone. Raqeeb reads the same Arabic briefing, from the same instructions,
and calls the same tool to record their decision. Only OpenAI minutes are spent.

This server never carries that audio. It mints a short-lived client secret for the session (the
API key must not reach a browser), and takes back the decision and the transcript when the
conversation ends.
"""

from __future__ import annotations

from typing import Literal

import httpx
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from agent.config import settings
from agent.graph.runner import get_incident_snapshot, resume_incident
from agent.providers.openai_llm import browser_audio_config, function_tools
from agent.voice.authority_prompts import RECORD_DISPATCH_CONFIRMATION_TOOL, build_dispatch_instructions

router = APIRouter(prefix="/api/incidents", tags=["browser-call"])

_CLIENT_SECRETS_URL = "https://api.openai.com/v1/realtime/client_secrets"

# A client secret is good for long enough to set the call up, not to hold one open: the
# conversation's own ceiling is settings.public_call_seconds, enforced by the browser.
_SECRET_SECONDS = 600

# One conversation per incident, so a visitor cannot reopen the same call over and over on our
# OpenAI account. Running a new detection is free to do and opens a new incident. Kept in memory:
# the deployment runs a single container, and a restart losing this only allows one extra
# conversation on an incident that survived it.
_started: set[str] = set()


class TranscriptLine(BaseModel):
    role: Literal["assistant", "authority"]
    text: str = Field(max_length=2000)


class BrowserCallResult(BaseModel):
    """What the browser saw. Anyone can post this, so nothing here is taken on trust beyond the
    incident it names: a confirmation is only recorded if the authority actually spoke, the same
    rule the phone path applies in agent/voice/sip_authority_call.py."""

    outcome: Literal["answered", "failed"] = "answered"
    dispatch_confirmed: bool = False
    authority_statement: str = Field(default="", max_length=2000)
    transcript: list[TranscriptLine] = Field(default_factory=list, max_length=200)


async def _awaiting_browser_call(incident_id: str):
    snapshot = await get_incident_snapshot(incident_id)
    if snapshot.next != ("gemini_authority_conversation",) or snapshot.values.get("call_transport") != "browser":
        raise HTTPException(status_code=409, detail="This incident is not waiting for a browser conversation.")
    return snapshot


@router.post("/{incident_id}/browser-call/token")
async def browser_call_token(incident_id: str):
    """A client secret for this incident's dispatch conversation, carrying its instructions.

    The session is fixed here rather than by the browser: the model, the voice, the briefing and
    the single tool it may call all come from the incident, so the secret cannot be repurposed
    into a general-purpose assistant on our account."""
    snapshot = await _awaiting_browser_call(incident_id)
    if incident_id in _started:
        raise HTTPException(status_code=409, detail="The conversation for this incident has already been held.")

    values = snapshot.values
    session = {
        "type": "realtime",
        "model": settings.openai_realtime_model,
        "instructions": build_dispatch_instructions(incident_id, values.get("report", {}), values.get("authority", {})),
        "audio": browser_audio_config(),
        "tools": function_tools([RECORD_DISPATCH_CONFIRMATION_TOOL]),
        "tool_choice": "auto",
    }
    async with httpx.AsyncClient(timeout=20) as client:
        response = await client.post(
            _CLIENT_SECRETS_URL,
            headers={"Authorization": f"Bearer {settings.openai_api_key}", "Content-Type": "application/json"},
            json={"session": session, "expires_after": {"anchor": "created_at", "seconds": _SECRET_SECONDS}},
        )
    if response.status_code >= 400:
        print(f"[browser-call {incident_id}] client secret refused: {response.status_code} {response.text[:300]}")
        raise HTTPException(status_code=502, detail="The conversation could not be started.")

    _started.add(incident_id)
    body = response.json()
    return {
        "client_secret": body["value"],
        "expires_at": body.get("expires_at"),
        "max_seconds": settings.public_call_seconds,
    }


@router.post("/{incident_id}/browser-call/result")
async def browser_call_result(incident_id: str, result: BrowserCallResult):
    """The end of the conversation: the authority's decision, or that it never happened."""
    await _awaiting_browser_call(incident_id)
    transcript = [line.model_dump() for line in result.transcript if line.text.strip()]
    heard_authority = any(line["role"] == "authority" for line in transcript)
    confirmed = bool(result.dispatch_confirmed and result.outcome == "answered" and heard_authority)

    await resume_incident(
        incident_id,
        {
            "dispatch_confirmed": confirmed,
            # "failed" holds the incident as one that reached no one, the same as a call that
            # never connected, so the dashboard explains itself the same way.
            "outcome": "answered" if result.outcome == "answered" else "failed",
            "authority_statement": result.authority_statement if confirmed else "",
            "raw_transcript": transcript,
        },
    )
    return {"incident_id": incident_id, "dispatch_confirmed": confirmed}
